broadinstitute/single_cell_portal_core
# Filtering most expression-based visualizations by cells appearing in matrix (SCP-2486)

Currently, Single Cell Portal only stores non-zero expression values when ingesting a dense expression matrix.  All zeros are discarded along with their associated cell names for that particular gene, and the zeros are then inferred at render time by using the list of cells from the given cluster, and comparing that against the retrieved expression data.  If the cell name does not appear in the list, then its expression is inferred as `0`.  While this is very often true, it is also possible that the given cell was not observed at all in the expression matrix, in which case the real expression value is `null` or `NaN`.  This has the unintended consequence of skewing distribution-based plots by over-inflating the true amount of `0` expression values, causing median and quartile values to be lower than they normally would be.

This update adds filtering to most (but not all) expression-based plots to account for this.  Now, when a given gene/cluster/annotation is requested, we take the intersection of cells from the matrix and cluster, and only provide expression values for that intersection.  Cells not appearing in the matrix are discarded from distribution-based plots, and for normal expression scatter plots, the value is set to `null`.

So far, this covers all distribution-based plots, such as violin, dot plots, and heatmaps, as well as any multi-gene query where the user collapses by a summary statistic.  2D annotation-based scatter plots (for numeric annotations vs. gene expression) and correlation plots are currently not covered in this feature due to the difficulty in maintaining associations between coordinate, expression, and annotation data based on how the data is assembled for those visualizations.  It would be possible to do, but likely have larger performance impacts than the above scenarios, so it is being exempted for now.

SCREENSHOTS

Normal single-gene expression scatter:
![Screen Shot 2022-06-01 at 3 17 31 PM](https://user-images.githubusercontent.com/729968/171485322-22ff22b1-48bb-4bd8-ab0a-8f678aaef500.png)

Single-gene violin plot (3rd donor ID has been completed removed, compared to reference plot above):
![Screen Shot 2022-06-01 at 3 17 40 PM](https://user-images.githubusercontent.com/729968/171485345-678a598d-29a2-4bfa-a2a7-fb1d8394c718.png)

Multi-gene dot plot (3rd donor ID has been completely removed, compared to reference plot above):
![Screen Shot 2022-06-01 at 3 17 56 PM](https://user-images.githubusercontent.com/729968/171485438-c9d0b796-fc1c-41dd-8f13-66afa7ae63a1.png)

Multi-gene heatmap (3rd donor ID has been completed removed, compared to reference plot above):
![Screen Shot 2022-06-01 at 3 18 02 PM](https://user-images.githubusercontent.com/729968/171485467-07bf736b-b490-437b-a46f-e3d16f01a6c2.png)

MANUAL TESTING
1. Pull branch, boot as normal, and start DelayedJob workers
2. In a rails console session, run `SyntheticStudyPopulator.populate('exp_cell_filtering')`
3. Once example files have ingested, load the explore tab for the above study and select `donor_id` as the annotation
4. Query for the gene `Foo`, and confirm the expression-based scatter and violin plots match above screenshots
5. Additionally query for `Bar`, and confirm the dotplot and heatmap match above screenshots

This PR satisfies SCP-2486
